src/bp/ui-shared/src/telemetry/offlineServerFallback.ts

```diff
@@ -0,0 +1,49 @@
+import axios, { AxiosInstance } from 'axios'
+import { TelemetryEvent } from 'common/telemetry'
+import ms from 'ms'
+
+export const sendEventsToTelemetryServer = async (events: TelemetryEvent[]) => {
+  try {
+    await axios.post(
+      window.TELEMETRY_URL,
+      events.map(e => ({ ...e, source: 'client' }))
+    )
+    return true
+  } catch (err) {
+    console.error('Could not send the telemetry packages to the storage server', err)
+    return false
+  }
+}
+
+export const startOfflineTelemetryFallback = (api: AxiosInstance) => {
+  sendSavedEventsRecursive()
+    .then(async () => {
+      setInterval(async () => await sendSavedEventsRecursive(), ms('30m'))
+    })
+    .catch() // silently fail
+
+  async function sendSavedEventsRecursive() {
+    const events = await getSavedEvents()
+    if (!events.length) {
+      return // break recursivity
+    }
+    const success = await sendEventsToTelemetryServer(events)
+    await sendFeedback(events, success)
+
+    await sendSavedEventsRecursive()
+  }
+
+  async function getSavedEvents(): Promise<TelemetryEvent[]> {
+    try {
+      const { data } = await api.get(`/telemetry/events`)
+      return data
+    } catch (err) {
+      return []
+    }
+  }
+
+  async function sendFeedback(events: TelemetryEvent[], success: boolean): Promise<void> {
+    const payload = { events: events.map(e => e.uuid), success }
+    return api.post(`/telemetry/feedback`, payload)
+  }
+}
```

Comment:
Always prefer arrow functions, unless you have a special need for `this` or special typings

Reply:
The functions are declared after they are used in this (made it this way to be easier to read) and using arrow functions makes them unusable before their declarations. Therefore, the normal function declaration is needed to make this special typing to work.

Reply:
You have 3 different methods inside one method, you can get them out of that method and use arrow functions.

Reply:
Fixed in 7e3743e!

Reply:
Fixed in 7e3743e!